src/main/java/frc/robot/swerve/SwerveDrive.java

```diff
@@ -0,0 +1,322 @@
+        }
+        return connected;
+    }
+
+    public Command checkSwerve() {
+        Command command = Commands.none();
+        for (SwerveModule module : modules) {
+            command = command.alongWith(module.checkModule());
+        }
+        return command;
+    }
+
+    public void stop() {
+        for (int i = 0; i < 4; i++) {
+            modules[i].setModuleState(new SwerveModuleState(0, modules[i].getModuleState().angle));
+        }
+    }
+
+    public void lock() {
+        modules[0].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(45)));
+        modules[1].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(135)));
+        modules[2].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(315)));
+        modules[3].setModuleState(new SwerveModuleState(0, Rotation2d.fromDegrees(225)));
+    }
+
+    @AutoLogOutput
+    public SwerveModulePosition[] getModulePositions() {
+        return modulePositions;
+    }
+
+    public void updateOffsets(double[] offsets) {
+        for (int i = 0; i < modules.length; i++) {
+            modules[i].updateOffset(offsets[i]);
+        }
+    }
+
+    /**
+     * Sets the correct module states from desired chassis speeds.
+     *
+     * @param chassisSpeeds Desired chassis speeds.
+     * @param fieldOriented Should the drive be field oriented.
+     */
+    public void drive(ChassisSpeeds chassisSpeeds, boolean fieldOriented) {
+        loggerInputs.desiredSpeeds = Utils.chassisSpeedsToArray(chassisSpeeds);
+
+        ChassisSpeeds fieldOrientedChassisSpeeds =
+                ChassisSpeeds.fromFieldRelativeSpeeds(
+                        chassisSpeeds.vxMetersPerSecond,
+                        chassisSpeeds.vyMetersPerSecond,
+                        chassisSpeeds.omegaRadiansPerSecond,
+                        new Rotation2d(getYaw()));
+
+        if (new ChassisSpeeds(0, 0, 0).equals(chassisSpeeds)) {
+            for (SwerveModule module : modules) {
+                module.stopMotor();
+            }
+        }
+
+        if (fieldOriented) {
+            setModuleStates(kinematics.toSwerveModuleStates(fieldOrientedChassisSpeeds));
+        } else {
+            setModuleStates(kinematics.toSwerveModuleStates(chassisSpeeds));
+        }
+    }
+
+    /**
+     * Sets the desired percentage of x, y and omega speeds for the frc.robot.swerve
+     *
+     * @param xOutput percentage of the x speed
+     * @param yOutput percentage of the y speed
+     * @param omegaOutput percentage of the omega speed
+     */
+    public void drive(double xOutput, double yOutput, double omegaOutput, boolean fieldOriented) {
+        ChassisSpeeds chassisSpeeds =
+                new ChassisSpeeds(
+                        SwerveConstants.MAX_X_Y_VELOCITY * xOutput,
+                        SwerveConstants.MAX_X_Y_VELOCITY * yOutput,
+                        SwerveConstants.MAX_OMEGA_VELOCITY * omegaOutput); // removed angleFF
```

Comment:
What does "removed angleFF" mean? Should probably be removed.

Reply:
08616497e7128678c8265215d8d0e215eac12ea7